File: lambdas/chat_proxy/handler.py
from __future__ import annotations

import base64
import json
import logging
import os
from typing import Any, Protocol, cast
from uuid import UUID

import boto3

MAX_MESSAGE_CHARS = 8_000
RUNTIME_QUALIFIER = "preview"
logger = logging.getLogger(__name__)


class RuntimeResponseBody(Protocol):
    def read(self) -> bytes: ...


class RuntimeClient(Protocol):
    def invoke_agent_runtime(self, **kwargs: object) -> dict[str, object]: ...

    def stop_runtime_session(self, **kwargs: object) -> object: ...


class ChatProxy:
    def __init__(self, client: RuntimeClient, runtime_arn: str) -> None:
        self._client = client
        self._runtime_arn = runtime_arn

    def handle(self, event: dict[str, Any], _context: object) -> dict[str, object]:
        method = event.get("httpMethod")
        path = event.get("path")
        if method == "OPTIONS" and path in {"/api/config", "/api/chat", "/api/reset"}:
            return _response(204, {})
        if method == "GET" and path == "/api/config":
            return _response(
                200,
                {"chatEnabled": True, "maxMessageChars": 8000, "maxTurns": 5},
            )
        if method == "POST" and path in {"/api/chat", "/api/reset"}:
            body = self._body(event)
            if body is None or not _valid_session(body.get("session_id")):
                return _invalid_request()
            if path == "/api/reset":
                return self._reset(body["session_id"])
            return self._chat(body)
        return _response(404, {"error": {"code": "not_found"}})

    @staticmethod
    def _body(event: dict[str, Any]) -> dict[str, Any] | None:
        raw = event.get("body")
        if not isinstance(raw, str) or len(raw) > MAX_MESSAGE_CHARS + 512:
            return None
        try:
            if event.get("isBase64Encoded"):
                raw = base64.b64decode(raw, validate=True).decode()
            value = cast(object, json.loads(raw))
        except (ValueError, UnicodeDecodeError):
            return None
        return cast(dict[str, Any], value) if isinstance(value, dict) else None

    def _chat(self, body: dict[str, Any]) -> dict[str, object]:
        message = body.get("message")
        if not isinstance(message, str) or not message.strip():
            return _invalid_request()
        message = message.strip()
        if len(message) > MAX_MESSAGE_CHARS:
            return _invalid_request()
        try:
            result = self._client.invoke_agent_runtime(
                agentRuntimeArn=self._runtime_arn,
                runtimeSessionId=body["session_id"],
                qualifier=RUNTIME_QUALIFIER,
                payload=json.dumps({"prompt": message}).encode(),
            )
            response_body = cast(RuntimeResponseBody, result["response"])
            runtime_response = cast(object, json.loads(response_body.read()))
            if not isinstance(runtime_response, dict):
                raise ValueError("AgentCore returned a non-object response")
            runtime_response = cast(dict[str, object], runtime_response)
            if "error" in runtime_response:
                return _response(422, runtime_response)
            answer = runtime_response.get("response")
            if not isinstance(answer, str):
                raise ValueError("AgentCore returned an invalid response")
            return _response(
                200,
                {
                    "answer": answer,
                    "session_id": body["session_id"],
                    "blocked": bool(runtime_response.get("blocked", False)),
                },
            )
        except Exception:
            logger.exception("AgentCore invocation failed")
            return _response(
                502,
                {
                    "error": {
                        "code": "agent_unavailable",
                        "message": "TollChat is temporarily unavailable. Please try again.",
                    }
                },
            )

    def _reset(self, session_id: str) -> dict[str, object]:
        try:
            self._client.stop_runtime_session(
                agentRuntimeArn=self._runtime_arn,
                runtimeSessionId=session_id,
                qualifier=RUNTIME_QUALIFIER,
            )
            return _response(200, {"ok": True})
        except Exception:
            logger.exception("AgentCore session reset failed")
            return _response(
                502,
                {
                    "error": {
                        "code": "agent_unavailable",
                        "message": "TollChat is temporarily unavailable. Please try again.",
                    }
                },
            )


def _valid_session(value: object) -> bool:
    if not isinstance(value, str):
        return False
    try:
        return str(UUID(value)) == value
    except ValueError:
        return False


def _invalid_request() -> dict[str, object]:
    return _response(
        400,
        {
            "error": {
                "code": "invalid_request",
                "message": "Provide a valid chat session and message.",
            }
        },
    )


def _response(status: int, body: dict[str, object]) -> dict[str, object]:
    return {
        "statusCode": status,
        "headers": {
            "Content-Type": "application/json",
            "Cache-Control": "no-store",
            "Access-Control-Allow-Origin": "https://tollchat.ai",
            "Access-Control-Allow-Headers": "content-type",
            "Access-Control-Allow-Methods": "GET,POST,OPTIONS",
        },
        "body": json.dumps(body),
    }


_proxy: ChatProxy | None = None


def lambda_handler(event: dict[str, Any], context: object) -> dict[str, object]:
    global _proxy
    if _proxy is None:
        client = cast(
            RuntimeClient,
            boto3.client(  # pyright: ignore[reportUnknownMemberType]
                "bedrock-agentcore",
                endpoint_url=os.environ.get("AGENTCORE_VPCE_URL"),
            ),
        )
        _proxy = ChatProxy(client, os.environ["AGENTCORE_RUNTIME_ARN"])
    return _proxy.handle(event, context)

File: lambdas/chat_proxy/handler.mjs
import { readFileSync } from "node:fs";
import {
  BedrockAgentCoreClient,
  InvokeAgentRuntimeCommand,
  StopRuntimeSessionCommand,
} from "@aws-sdk/client-bedrock-agentcore";

const MAX_MESSAGE_CHARS = 8_000;
const SAFE_ERROR = {
  type: "error",
  code: "agent_unavailable",
  message: "TollChat is temporarily unavailable. Please try again.",
};
const LABELS = new Set([
  "Planning toll route",
  "Checking I-66 tolls",
  "Checking I-95/395 tolls",
  "Checking I-95/395 Express Lanes access",
  "Checking I-95/395 Express Lanes tolls",
  "Checking I-495 tolls",
  "Checking Dulles tolls",
  "Checking toll data",
]);
const STATUS = new Set(["running", "completed", "failed"]);
const ERROR_CODES = new Set(["invalid_request", "turn_limit", "agent_unavailable"]);
const ERROR_MESSAGES = new Set([
  "Provide a message between 1 and 8000 characters.",
  "Start a new chat to continue.",
  "TollChat could not complete that request. Please try again.",
]);

const json = (statusCode, value) => ({
  statusCode,
  headers: { "Content-Type": "application/json", "Cache-Control": "no-store" },
  body: JSON.stringify(value),
});

const invalid = () => json(400, {
  error: { code: "invalid_request", message: "Provide a valid chat session and message." },
});

const parseBody = (event) => {
  if (typeof event.body !== "string" || event.body.length > MAX_MESSAGE_CHARS + 512) return null;
  try {
    const raw = event.isBase64Encoded
      ? Buffer.from(event.body, "base64").toString("utf8")
      : event.body;
    const value = JSON.parse(raw);
    return value && typeof value === "object" && !Array.isArray(value) ? value : null;
  } catch {
    return null;
  }
};

const validSession = (value) =>
  typeof value === "string" && /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(value);

const exactKeys = (value, keys) =>
  Object.keys(value).sort().join(",") === [...keys].sort().join(",");

const validEvent = (value) => {
  if (!value || typeof value !== "object" || Array.isArray(value)) return false;
  if (value.type === "tool") {
    return exactKeys(value, ["type", "index", "label", "status"])
      && Number.isInteger(value.index) && value.index >= 0
      && LABELS.has(value.label) && STATUS.has(value.status);
  }
  if (value.type === "answer") {
    return exactKeys(value, ["type", "text", "blocked"])
      && typeof value.text === "string" && typeof value.blocked === "boolean";
  }
  return value.type === "error"
    && exactKeys(value, ["type", "code", "message"])
    && ERROR_CODES.has(value.code) && ERROR_MESSAGES.has(value.message);
};

async function* ndjsonFromSse(stream) {
  const decoder = new TextDecoder();
  let buffer = "";
  let terminal;
  try {
    for await (const chunk of stream) {
      buffer += decoder.decode(chunk, { stream: true }).replaceAll("\r", "");
      let boundary;
      while ((boundary = buffer.indexOf("\n\n")) >= 0) {
        const frame = buffer.slice(0, boundary);
        buffer = buffer.slice(boundary + 2);
        if (!frame || frame.startsWith(":")) continue;
        const lines = frame.split("\n");
        if (lines.some((line) => !line.startsWith("data: "))) throw new Error("invalid SSE frame");
        const value = JSON.parse(lines.map((line) => line.slice(6)).join("\n"));
        if (!validEvent(value)) throw new Error("invalid stream event");
        if (terminal) throw new Error("event after terminal");
        if (value.type === "answer" || value.type === "error") terminal = value;
        else yield `${JSON.stringify(value)}\n`;
      }
    }
    if (buffer.trim()) throw new Error("incomplete SSE frame");
    if (!terminal) throw new Error("missing terminal event");
    yield `${JSON.stringify(terminal)}\n`;
  } catch {
    yield `${JSON.stringify(SAFE_ERROR)}\n`;
  }
}

export async function route(event, dependencies) {
  const { client, runtimeArn, previewHtml, previewJs = "", previewAssets = {} } = dependencies;
  const method = event.httpMethod;
  const path = event.path;
  if (method === "GET" && path === "/") {
    return { statusCode: 200, headers: { "Content-Type": "text/html; charset=utf-8", "Cache-Control": "no-store" }, body: previewHtml };
  }
  if (method === "GET" && path === "/preview.mjs") {
    return { statusCode: 200, headers: { "Content-Type": "text/javascript; charset=utf-8", "Cache-Control": "no-store" }, body: previewJs };
  }
  if (method === "GET" && Object.hasOwn(previewAssets, path)) {
    return { statusCode: 200, headers: { "Content-Type": "text/javascript; charset=utf-8", "Cache-Control": "public, max-age=31536000, immutable" }, body: previewAssets[path] };
  }
  if (method === "GET" && path === "/api/config") {
    return json(200, { chatEnabled: true, maxMessageChars: MAX_MESSAGE_CHARS, maxTurns: 5 });
  }
  if (method !== "POST" || !["/api/chat", "/api/reset"].includes(path)) {
    return json(404, { error: { code: "not_found" } });
  }
  const body = parseBody(event);
  if (!body || !validSession(body.session_id)) return invalid();
  try {
    if (path === "/api/reset") {
      await client.send(new StopRuntimeSessionCommand({
        agentRuntimeArn: runtimeArn,
        runtimeSessionId: body.session_id,
        qualifier: "preview",
      }));
      return json(200, { ok: true });
    }
    if (typeof body.message !== "string" || !body.message.trim() || body.message.trim().length > MAX_MESSAGE_CHARS) return invalid();
    const result = await client.send(new InvokeAgentRuntimeCommand({
      agentRuntimeArn: runtimeArn,
      runtimeSessionId: body.session_id,
      qualifier: "preview",
      payload: new TextEncoder().encode(JSON.stringify({ prompt: body.message.trim() })),
    }));
    if (!result.contentType?.includes("text/event-stream") || !result.response?.[Symbol.asyncIterator]) throw new Error("invalid upstream response");
    return {
      statusCode: 200,
      headers: { "Content-Type": "application/x-ndjson", "Cache-Control": "no-store", "X-Content-Type-Options": "nosniff" },
      body: ndjsonFromSse(result.response),
    };
  } catch (error) {
    console.error("AgentCore request failed", error?.name ?? "Error");
    return json(502, { error: SAFE_ERROR });
  }
}

const client = new BedrockAgentCoreClient({
  region: process.env.AWS_REGION,
  endpoint: process.env.AGENTCORE_VPCE_URL,
});
const deployed = Boolean(process.env.AWS_LAMBDA_RUNTIME_API);
const dependencies = deployed ? {
  client,
  runtimeArn: process.env.AGENTCORE_RUNTIME_ARN,
  previewHtml: readFileSync(new URL("./preview.html", import.meta.url), "utf8"),
  previewJs: readFileSync(new URL("./preview.mjs", import.meta.url), "utf8"),
  previewAssets: {
    "/assets/chat-markdown-v1.mjs": readFileSync(new URL("./assets/chat-markdown-v1.mjs", import.meta.url), "utf8"),
    "/assets/markdown-it-15.0.0/markdown-it.esm.min.mjs": readFileSync(new URL("./assets/markdown-it-15.0.0/markdown-it.esm.min.mjs", import.meta.url), "utf8"),
  },
} : {};

export const handler = deployed ? globalThis.awslambda.streamifyResponse(async (event, responseStream) => {
  const response = await route(event, dependencies);
  const stream = globalThis.awslambda.HttpResponseStream.from(responseStream, {
    statusCode: response.statusCode,
    headers: response.headers,
  });
  if (typeof response.body === "string") stream.write(response.body);
  else for await (const chunk of response.body) stream.write(chunk);
  stream.end();
}) : undefined;
